core/src/main/java/hudson/tasks/Fingerprinter.java

```diff
@@ -92,13 +95,40 @@
      */
     private final String targets;
 
+    /**
+     * Default null 'excludes' pattern as in Ant.
+     */
+    private String excludes = null;
+
+    /**
+     * Default ant exclusion
+     */
+    private boolean defaultExcludes = true;
+
+    /**
+     * Indicate whether include and exclude patterns should be considered as case sensitive
+     */
+    private boolean caseSensitive = true;
```

Comment:
You need to be careful with field initializers which do not match the Java defaults. For Pipeline this is no problem because every `fingerprint` step call will call the `@DataBoundConstructor` and so these initializer expressions will be applied. But for steps in freestyle (etc.) projects, where the `Fingerprinter` object was serialized to `config.xml`, after an upgrade the initializer expressions will be ignored for existing steps—so while `excludes` will be `null` as expected, `defaultExcludes` and `caseSensitive` will both be `false`.

This can be addressed with a `readResolve` if you switch the latter two fields to be of type `Boolean` so you can distinguish unconfigured from old installation (`null`) from deliberately off (`false` ~ `FALSE`).

(You can also invert the sense of the fields while retaining the sense of the getters and setters using `!`, though this is arguably more confusing than using `Boolean`.)

Reply:
FWIU, `Boolean` implements `Serializable` interface which in turn solves our serialization issue. Am I right, here?

Reply:
That's unrelated, Xstream just serializes everything.

The problem is that if you introduce a new field, and an object gets deserialized from a previous Jenkins version that did not have the field, it's `null` or the default value for a primitive. No constructors get called and nothing. So this will start out being `false` when someone updated their Jenkins.

Reply:
Will Xstream at least try to use a registered Converter? Otherwise, how would you migrate String to Secret without additional support code?

Reply:
@jglick Thanks for explanation.
@daniel-beck Thanks for clearing out doubt.